cf-custom-resources/lib/cert-replicator.js

```diff
@@ -0,0 +1,324 @@
+ * If the certificate does not exist, the function will return normally.
+ *
+ * @param {string} arn The certificate ARN
+ * @param {string} targetRegionAcm acm client in target region
+ */
+const deleteCertificate = async function (arn, acm) {
+  try {
+    console.log(`Waiting for certificate ${arn} to become unused`);
+
+    let inUseByResources;
+    let options;
+
+    for (let attempt = 0; attempt < maxAttempts; attempt++) {
+      const { Certificate } = await acm
+        .describeCertificate({
+          CertificateArn: arn,
+        })
+        .promise();
+
+      inUseByResources = Certificate.InUseBy || [];
+      options = Certificate.DomainValidationOptions || [];
+      let ok = false;
+      for (const option of options) {
+        if (!option.ResourceRecord) {
+          ok = false;
+          break;
+        }
+        ok = true;
+      }
+      if (!ok || inUseByResources.length) {
+        // Deleting resources can be quite slow - so just sleep 30 seconds between checks.
+        await sleep(30000);
+      } else {
+        break;
+      }
+    }
+    if (inUseByResources.length) {
+      throw new Error(
+        `Certificate still in use after checking for ${maxAttempts} attempts.`
+      );
+    }
+
+    await acm
+      .deleteCertificate({
+        CertificateArn: arn,
+      })
+      .promise();
+  } catch (err) {
+    if (err.name !== "ResourceNotFoundException") {
+      throw err;
+    }
+  }
+};
+
+const validateCertificate = async function (certificateARN, acm) {
+  await acm
+    .waitFor("certificateValidated", {
+      // Wait up to 9 minutes and 30 seconds
+      $waiter: {
+        delay: 30,
+        maxAttempts: 19,
+      },
+      CertificateArn: certificateARN,
+    })
+    .promise();
+};
+
+const acmClient = function (region) {
+  const acm = new aws.ACM({
+    region,
+  });
+  if (waiter) {
+    // Used by the test suite, since waiters aren't mockable yet
+    acm.waitFor = waiter;
+  }
+  return acm;
+};
+
+/**
+ * Main certificate replicator handler, invoked by Lambda
+ */
+exports.certificateReplicateHandler = async function (event, context) {
+  let responseData = {};
+  let physicalResourceId = event.PhysicalResourceId;
+  const props = event.ResourceProperties;
+  const [targetRegion, envRegion, certArn] = [
+    props.TargetRegion,
+    props.EnvRegion,
+    props.CertificateArn,
+  ];
+
+  const envRegionAcm = acmClient(envRegion);
+  const targetRegionAcm = acmClient(targetRegion);
+
+  try {
+    let response = {};
+    switch (event.RequestType) {
+      case "Create":
+      case "Update":
+        response = await replicateCertificate(
+          event.RequestId,
+          props.AppName,
+          props.EnvName,
+          certArn,
+          envRegionAcm,
+          targetRegionAcm
+        );
+        responseData.Arn = physicalResourceId = response.CertificateArn;
+        await validateCertificate(response.CertificateArn, targetRegionAcm);
```

Comment:
I'm trying to figure out if setting the physical resource ID before validating the certificate is the right thing to do:

1. It's an `Update` event.
2.  ✅  `replicateCertificate` is successfully and we update the `physicalResourceId` successfully 
3. ❌ `validateCertificate` then fails for some reason.
4. We go into the `catch` block with the **new** physical id.

In this scenario, _I think_ on rollback we will delete the new cert and the rollback will complete 👍 Can you validate if this is right?

Reply:
Yeah that's exactly the reason why we set the `physicalResourceId` early. Otherwise even if `validateCertificate` fails, the new cert won't be removed by CFN